drivers/sensor/hx711/hx711.c

```diff
@@ -0,0 +1,584 @@
+ * @param offset sensor_value struct. Offset used to calculate the real weight.
+ *        Formula is :
+ *          weight = slope * (reading - offset)
+ *          Only val1 is used.
+ *
+ * @return void
+ *
+ */
+static void hx711_attr_set_offset(struct hx711_data *data, const struct sensor_value *offset)
+{
+	data->offset = offset->val1;
+}
+
+/**
+ * @brief Set HX711 slope.
+ *
+ * @param data Pointer to the hx711 driver data structure
+ * @param slope sensor_value struct. Slope used to calculate the real weight.
+ *        Formula is :
+ *          weight = slope * (reading - offset)
+ *
+ * @return void
+ *
+ */
+static void hx711_attr_set_slope(struct hx711_data *data, const struct sensor_value *slope)
+{
+	data->slope.val1 = slope->val1;
+	data->slope.val2 = slope->val2;
+}
+
+/**
+ * @brief Set HX711 attributes.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param chan Channel to set.
+ *             Supported channels :
+ *               SENSOR_CHAN_WEIGHT
+ *               SENSOR_CHAN_ALL
+ * @param attr Attribute to change.
+ *             Supported attributes :
+ *               SENSOR_ATTR_SAMPLING_FREQUENCY
+ *               SENSOR_ATTR_OFFSET
+ *               SENSOR_ATTR_SLOPE
+ *               SENSOR_ATTR_GAIN
+ * @param val   Value to set.
+ * @retval 0 on success
+ * @retval -ENOTSUP if an invalid attribute is given
+ *
+ */
+static int hx711_attr_set(const struct device *dev, enum sensor_channel chan,
+			  enum sensor_attribute attr, const struct sensor_value *val)
+{
+	int ret = 0;
+	struct hx711_data *data = dev->data;
+
+	switch (attr) {
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		ret = hx711_attr_set_rate(dev, val);
+		if (ret == 0) {
+			LOG_DBG("Attribute RATE set to %d", data->rate);
+		}
+		return ret;
+	case SENSOR_ATTR_OFFSET:
+		hx711_attr_set_offset(data, val);
+		LOG_DBG("Attribute OFFSET set to %d", data->offset);
+		return ret;
+	case SENSOR_ATTR_SLOPE:
+		hx711_attr_set_slope(data, val);
+		LOG_DBG("Attribute SLOPE set to %d.%d", data->slope.val1, data->slope.val2);
+		return ret;
+	case SENSOR_ATTR_GAIN:
+		ret = hx711_attr_set_gain(dev, val);
+		if (ret == 0) {
+			LOG_DBG("Attribute GAIN set to %d", data->gain);
+		}
+		return ret;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+/**
+ * @brief Get HX711 reading.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param chan Channel to set.
+ *             Supported channels :
+ *               SENSOR_CHAN_WEIGHT
+ *               SENSOR_CHAN_ALL
+ *
+ * @param val  Value to write weight value to.
+ *        Formula is :
+ *          weight = slope * (reading - offset)
+ *
+ * @retval 0 on success
+ * @retval  -ENOTSUP if an invalid channel is given
+ *
+ */
+static int hx711_channel_get(const struct device *dev, enum sensor_channel chan,
+			     struct sensor_value *val)
+{
+	struct hx711_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_WEIGHT:
+	case SENSOR_CHAN_ALL: {
```

Comment:
I looked it up in the documentation I found it mentioned that **sensor_channel_get** should 
"..Return a useful value for a **particular** channel .."
[sensor_channel_get](https://docs.zephyrproject.org/latest/reference/peripherals/sensor.html?highlight=channel_get#c.sensor_channel_get).

Since the function signature is : 
_int sensor_channel_get(const struct device *dev, **enum sensor_channel chan**, struct sensor_value *val)_
and **SENSOR_CHAN_ALL** is of type **enum sensor_channel** then **SENSOR_CHAN_ALL** is accepted as a value.

The current function implementation accepts both and rejects any other invalid channel.

**SENSOR_CHAN_WEIGHT** is also subset of **SENSOR_CHAN_ALL** and in this particular case also equal.
It would seem strange from the user perspective to ask for all channels on a sensor which has only one and get an error instead of a valid **sensor_value**.

Of course it would be an issue for sensors with multiple channels. The API should then return a pointer to an sensor_value array or something which is not currently the case.

I understand your argument though.
It _should_ only be supported when fetching but is currently not ;)

Reply:
Removed case SENSOR_CHAN_ALL
Done.